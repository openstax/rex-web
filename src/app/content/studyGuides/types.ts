import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import {
  CountsPerSource,
  SummaryHighlights,
  SummaryHighlightsPagination,
} from '../highlights/types';

export interface StudyGuidesSummaryFilters {
  colors: HighlightColorEnum[];
  locationIds: string[];
}

export interface State {
  isEnabled: boolean;
  summary: {
    filters: StudyGuidesSummaryFilters,
    loading: boolean,
    open: boolean,
    pagination: SummaryHighlightsPagination,
    studyGuides: SummaryHighlights | null
    totalCountsPerPage: CountsPerSource | null;
  };
}
