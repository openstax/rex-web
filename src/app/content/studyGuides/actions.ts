import { createStandardAction } from 'typesafe-actions';
import {
  CountsPerSource,
  SummaryFilters,
  SummaryFiltersUpdate,
  SummaryHighlights,
  SummaryHighlightsPagination
} from '../highlights/types';
import { StudyGuidesSummaryFilters } from './types';

export const receiveSummaryStudyGuides = createStandardAction(
  'Content/StudyGuides/Summary/receive'
)<
SummaryHighlights,
  {
    pagination: SummaryHighlightsPagination,
    filters?: StudyGuidesSummaryFilters,
    isStillLoading?: boolean
  }
>();

export const openStudyGuides = createStandardAction('Content/StudyGuides/Summary/open')<void>();
export const closeStudyGuides = createStandardAction('Content/StudyGuides/Summary/close')<void>();
export const loadMoreStudyGuides = createStandardAction('Content/StudyGuides/loadMore')();
export const printStudyGuides = createStandardAction('Content/StudyGuides/print')<void>();
export const setDefaultSummaryFilters = createStandardAction('Content/StudyGuides/Summary/setDefaultFilters')<
  SummaryFilters
>();
export const setSummaryFilters = createStandardAction('Content/StudyGuides/Summary/setFilters')<
  Partial<SummaryFilters>
>();
export const updateSummaryFilters = createStandardAction('Content/StudyGuides/Summary/updateFilters')<
  SummaryFiltersUpdate
>();
export const toggleStudyGuidesSummaryLoading = createStandardAction('Content/StudyGuides/Summary/loading')<boolean>();
export const receiveStudyGuidesTotalCounts = createStandardAction(
  'Content/StudyGuides/receiveTotalCounts'
)<CountsPerSource>();
