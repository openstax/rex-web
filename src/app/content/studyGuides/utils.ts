import { HighlightColorEnum } from '@openstax/highlights-client';
import { OutputParams } from 'query-string';
import { colorFilterQueryParameterName, locationIdsFilterQueryParameterName } from '../constants';
import { colorfilterLabels } from './constants';

export const getFiltersFromQuery = (query: OutputParams) => {
  const queryColors = query[colorFilterQueryParameterName] as HighlightColorEnum | HighlightColorEnum[] | undefined;
  const queryLocationIds = query[locationIdsFilterQueryParameterName] as string | string[] | undefined;

  const colors = (Array.isArray(queryColors) ? queryColors : [queryColors])
    .filter((color) => color && colorfilterLabels.has(color)) as HighlightColorEnum[];

  const locationIds = (Array.isArray(queryLocationIds) ? queryLocationIds : [queryLocationIds])
    .filter((id) => id) as string[];

  return { colors, locationIds };
};
