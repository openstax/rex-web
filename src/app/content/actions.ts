import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as practiceQuestionsActions from './practiceQuestions/actions';
import * as searchActions from './search/actions';
import * as studyGuidesActions from './studyGuides/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;
export const studyGuides = studyGuidesActions;
export const practiceQuestions = practiceQuestionsActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const receiveBuyPrintConfig = createStandardAction('Content/receiveBuyPrintConfig')<State['buyPrint']>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();
export const receivePageNotFoundId = createStandardAction('Content/receivePageNotFoundId')<string>();

export const openNudgeStudyTools = createStandardAction('Content/openNudgeStudyTools')();
export const closeNudgeStudyTools = createStandardAction('Content/closeNudgeStudyTools')();
