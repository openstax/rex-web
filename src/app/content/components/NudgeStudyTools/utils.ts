import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import { useDebouncedWindowSize, useOnScrollTopOffset } from '../../../reactUtils';
import { assertDocument, remsToPx } from '../../../utils';
import { page as pageSelector } from '../../selectors';
import { hasStudyGuides } from '../../studyGuides/selectors';
import { Page } from '../../types';
import {
  arrowDesktopHeight,
  arrowRightMargin,
  arrowTopMargin,
  closeButtonDistanceFromContent,
  closeButtonSize,
  contentMarginLeft,
  contentMarginTop,
  contentWidth,
  cookieNudge,
  daysUntilCookieExpires,
  mobileNudgeStudyToolsTargetId,
  mobileSpotlightPadding,
  nudgeStudyToolsMinPageLimit,
  nudgeStudyToolsShowLimit,
  nudgeStudyToolsTargetId,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';

interface Positions {
  arrowLeft: number;
  arrowTopOffset: number;
  closeButtonLeft: number;
  closeButtonTopOffset: number;
  contentWrapperLeft: number;
  contentWrapperTopOffset: number;
  spotlightHeight: number;
  spotlightLeftOffset: number;
  spotlightTopOffset: number;
  spotlightWidth: number;
}

export const getPositions = (target: HTMLElement, isMobile: boolean): Positions => {
  const padding = remsToPx(mobileSpotlightPadding);
  const { top, left, height, width } = getBoundingRectOfNudgeTarget(target);
  const spotlightLeftOffset = left - (isMobile ? padding : 0);
  const spotlightWidth = width + (isMobile ? padding * 2 : 0);
  const arrowTopOffset = top + height + remsToPx(arrowTopMargin);
  // left edge of arrow image should be on the middle of the spotlight (adjusted for a margin)
  const arrowLeft = spotlightLeftOffset + (spotlightWidth / 2) - remsToPx(arrowRightMargin);

  const contentWrapperTopOffset = arrowTopOffset
    + remsToPx(arrowDesktopHeight)
    + remsToPx(contentMarginTop);
  const contentWrapperLeft = spotlightLeftOffset + spotlightWidth + remsToPx(contentMarginLeft);
  const closeButtonLeft = contentWrapperLeft + remsToPx(contentWidth) + remsToPx(closeButtonDistanceFromContent);
  const closeButtonTopOffset = contentWrapperTopOffset
    - remsToPx(closeButtonSize)
    - remsToPx(closeButtonDistanceFromContent);

  return {
    arrowLeft,
    arrowTopOffset,
    closeButtonLeft,
    closeButtonTopOffset,
    contentWrapperLeft,
    contentWrapperTopOffset,
    spotlightHeight: height,
    spotlightLeftOffset,
    spotlightTopOffset: top,
    spotlightWidth,
  };
};

// Target may have display set to `contents` so we are calculatings rect for children of passed target.
const getBoundingRectOfNudgeTarget = (target: HTMLElement) => {
  const { top, bottom, left, right } = Array.from(target.children).reduce((rects, child) => {
    const rect = child.getBoundingClientRect();
    rects.top = Math.min(rects.top, rect.top);
    rects.bottom = Math.max(rects.bottom, rect.bottom);
    rects.left = Math.min(rects.left, rect.left);
    rects.right = Math.max(rects.right, rect.right);
    return rects;
    // Starting values depends on if we checks for Math.max or Math.min
  }, { top: 9999, bottom: 0, left: 9999, right: 0 });

  return {
    height: bottom - top,
    left,
    right,
    top,
    width: right - left,
  };
};

export const usePositions = (isMobile: boolean) => {
  const [windowWidth] = useDebouncedWindowSize();
  const scrollTopOffset = useOnScrollTopOffset();
  const [positions, setPositions] = React.useState<Positions | null>(null);
  const studyGuides = useSelector(hasStudyGuides);

  React.useEffect(() => {
    const document = assertDocument();
    const target = document.querySelector(
      `#${isMobile ? mobileNudgeStudyToolsTargetId : nudgeStudyToolsTargetId}`
    ) as HTMLElement | null;
    if (target) {
      // Make sure that we calculate positions with body overflow set to hidden
      // because it causes scrollbar to hide which results in different positions.
      const prevOverflow = document.body.style.overflow;
      document.body.style.overflow = 'hidden';
      setPositions(getPositions(target, isMobile));

      // Resets to the value from before calculations. We want this style change to be handled
      // directly in the component.
      document.body.style.overflow = prevOverflow;
    }
    return () => setPositions(null);
  }, [scrollTopOffset, studyGuides, windowWidth, isMobile]);

  return positions;
};

export const getCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.counter) || 0);
};

export const getDateCookie = () => {
  const lastShownDate = Cookies.get(cookieNudge.date);
  return lastShownDate ? new Date(lastShownDate) : undefined;
};

export const passedTimeInterval = () => {
  const lastShownDate = getDateCookie();
  return !lastShownDate
    ? true
    : (Date.now() - lastShownDate.getTime()) > timeIntervalBetweenShowingNudgeInMs
  ;
};

export const getPageCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.pageCounter) || 0);
};

export const incrementPageCounterCookie = () => {
  const counter = getPageCounterCookie();
  const newValue = counter + 1;
  Cookies.set(cookieNudge.pageCounter, newValue.toString(), {expires: daysUntilCookieExpires});
  return newValue;
};

export const shouldDisplayNudgeStudyTools = (): boolean => {
  const counter = getCounterCookie();
  const numberOfPagesOpenedByUser = getPageCounterCookie();

  return counter < nudgeStudyToolsShowLimit
    && passedTimeInterval()
    && (numberOfPagesOpenedByUser >= nudgeStudyToolsMinPageLimit);
};

// Set required cookies and reset opened page counter
export const setNudgeStudyToolsCookies = () => {
  const now = new Date();
  const counter = getCounterCookie();
  Cookies.set(cookieNudge.counter, (counter + 1).toString(), {expires: daysUntilCookieExpires});
  Cookies.set(cookieNudge.date, now.toString(), {expires: daysUntilCookieExpires});
  Cookies.remove(cookieNudge.pageCounter);
};

export const useIncrementPageCounter = () => {
  const page = useSelector(pageSelector);
  const [counter, updateCounter] = useState<number>(getPageCounterCookie());
  const pageRef = React.useRef<Page>();

  React.useEffect(() => {
    if (page && page !== pageRef.current && counter <= nudgeStudyToolsMinPageLimit) {
      pageRef.current = page;
      updateCounter(incrementPageCounterCookie());
    }
  }, [page, updateCounter, counter]);

  return counter;
};
