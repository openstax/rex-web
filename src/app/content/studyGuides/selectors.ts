import { createSelector } from 'reselect';
import { getHighlightColorFiltersWithContent, getHighlightLocationFilterForPage } from '../highlights/utils';
import {
  getHighlightLocationFilters,
  getHighlightLocationFiltersWithContent,
  getSortedSummaryHighlights
} from '../highlights/utils';
import {
  checkIfHasMoreResults,
  filterCounts,
  getLoadedCountsPerSource,
  getSelectedHighlightsLocationFilters,
} from '../highlights/utils/selectorsUtils';
import * as parentSelectors from '../selectors';
import { archiveTreeSectionIsChapter } from '../utils/archiveTreeUtils';

export const localState = createSelector(
  parentSelectors.localState,
  (parentState) => parentState.studyGuides
);

export const studyGuidesEnabled = createSelector(
  localState,
  (state) => state.isEnabled
);

export const studyGuidesSummary = createSelector(
  localState,
  (state) => state.summary
);

export const hasStudyGuides = createSelector(
  studyGuidesSummary,
  (summary) => summary.totalCountsPerPage
    && Object.keys(summary.totalCountsPerPage).length > 0
);

export const totalCountsPerPage = createSelector(
  studyGuidesSummary,
  (summary) => summary.totalCountsPerPage
);

export const totalCountsPerPageOrEmpty = createSelector(
  studyGuidesSummary,
  (summary) => summary.totalCountsPerPage || {}
);

export const studyGuidesOpen = createSelector(
  studyGuidesSummary,
  (summary) => summary.open
);

export const summaryIsLoading = createSelector(
  studyGuidesSummary,
  (summary) => summary.loading
);

export const summaryStudyGuidesPagination = createSelector(
  studyGuidesSummary,
  (summary) => summary.pagination
);

export const studyGuidesLocationFilters = createSelector(
  parentSelectors.book,
  getHighlightLocationFilters(archiveTreeSectionIsChapter)
);

export const summaryStudyGuides = createSelector(
  studyGuidesSummary,
  (summary) => summary.studyGuides
);

export const orderedSummaryStudyGuides = createSelector(
  summaryStudyGuides,
  studyGuidesLocationFilters,
  getSortedSummaryHighlights
);

export const loadedCountsPerSource = createSelector(
  summaryStudyGuides,
  getLoadedCountsPerSource
);

export const summaryFilters = createSelector(
  localState,
  (state) => state.summary.filters
);

const rawSummaryLocationFilters = createSelector(
  summaryFilters,
  (filters) => filters.locationIds
);

export const highlightColorFiltersWithContent = createSelector(
  totalCountsPerPageOrEmpty,
  getHighlightColorFiltersWithContent
);

export const summaryColorFilters = createSelector(
  summaryFilters,
  highlightColorFiltersWithContent,
  (filters, withContent) => new Set(filters.colors.filter((color) => withContent.has(color)))
);

export const studyGuidesLocationFiltersWithContent = createSelector(
  studyGuidesLocationFilters,
  totalCountsPerPageOrEmpty,
  getHighlightLocationFiltersWithContent
);

export const defaultLocationFilter = createSelector(
  studyGuidesLocationFilters,
  parentSelectors.page,
  (filters, currentPage) => currentPage && getHighlightLocationFilterForPage(filters, currentPage)
);

export const summaryLocationFilters = createSelector(
  rawSummaryLocationFilters,
  (selectedLocations) => new Set(selectedLocations)
);

const selectedStudyGuidesLocationFilters = createSelector(
  studyGuidesLocationFilters,
  summaryLocationFilters,
  getSelectedHighlightsLocationFilters
);

export const filteredCountsPerPage = createSelector(
  totalCountsPerPageOrEmpty,
  selectedStudyGuidesLocationFilters,
  summaryColorFilters,
  filterCounts
);

export const hasMoreResults = createSelector(
  loadedCountsPerSource,
  filteredCountsPerPage,
  summaryStudyGuidesPagination,
  checkIfHasMoreResults
);
