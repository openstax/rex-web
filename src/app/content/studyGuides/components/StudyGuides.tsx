import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import { useSelector } from 'react-redux';
import styled from 'styled-components/macro';
import { typesetMath } from '../../../../helpers/mathjax';
import htmlMessage from '../../../components/htmlMessage';
import Loader from '../../../components/Loader';
import { useServices } from '../../../context/Services';
import theme from '../../../theme';
import { assertWindow } from '../../../utils';
import SectionHighlights, {
  HighlightsChapterWrapper,
  HighlightSection,
  HighlightWrapper
} from '../../components/SectionHighlights';
import allImagesLoaded from '../../components/utils/allImagesLoaded';
import { GeneralCenterText } from '../../highlights/components/HighlightStyles';
import HighlightsWrapper from '../../styles/HighlightsWrapper';
import LoaderWrapper from '../../styles/LoaderWrapper';
import * as selectors from '../selectors';
import StudyGuidesListElement from './StudyGuidesListElement';

// tslint:disable-next-line: variable-name
export const NoStudyGuidesTip = htmlMessage(
  'i18n:studyguides:popup:no-highlights-tip',
  (props) => <span {...props} />
);

// tslint:disable-next-line: variable-name
const StudyGuides = ({ className }: { className: string }) => {
  const orderedStudyGuides = useSelector(selectors.orderedSummaryStudyGuides);
  const isLoading = useSelector(selectors.summaryIsLoading);
  const container = React.useRef<HTMLElement>(null);
  const services = useServices();

  React.useLayoutEffect(() => {
    if (container.current) {
      services.promiseCollector.add(allImagesLoaded(container.current));
      services.promiseCollector.add(typesetMath(container.current, assertWindow()));
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [orderedStudyGuides]);

  return <div className={className}>
    {isLoading ? <LoaderWrapper><Loader large /></LoaderWrapper> : null}
    {(!isLoading && orderedStudyGuides && orderedStudyGuides.length === 0) ?
      <HighlightsWrapper ref={container}>
        <GeneralCenterText>
          <FormattedMessage id='i18n:studyguides:popup:no-highlights'>
            {(msg) => msg}
          </FormattedMessage>
          <NoStudyGuidesTip />
        </GeneralCenterText>
      </HighlightsWrapper>
    : orderedStudyGuides && <HighlightsWrapper ref={container}>
        {orderedStudyGuides.map((highlightData) => {
          return <SectionHighlights
            key={highlightData.location.id}
            highlightDataInSection={highlightData}
            highlightRenderer={(highlight) => (
              <StudyGuidesListElement
                key={highlight.id}
                highlight={highlight}
              />
            )}
          />;
        })}
      </HighlightsWrapper>}
  </div>;
};

export default styled(StudyGuides)`
  ${HighlightsChapterWrapper} {
    ${theme.breakpoints.mobile`
      max-width: 90%;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    `}
  }

  ${HighlightSection} {
    ${theme.breakpoints.mobile`
      padding-left: 2rem;
    `}
  }

  @media print {
    ${HighlightsChapterWrapper} {
      padding-left: 0;

      & + ${HighlightWrapper} {
        margin-top: 0;
      }
    }

    ${HighlightWrapper} {
      margin-left: 0;
    }

    ${HighlightSection} {
      background: ${theme.color.neutral.darkest};
    }
  }

  display: contents;
`;
