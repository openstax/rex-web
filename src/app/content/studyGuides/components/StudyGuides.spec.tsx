import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import React from 'react';
import renderer from 'react-test-renderer';
import { typesetMath } from '../../../../helpers/mathjax';
import createTestServices from '../../../../test/createTestServices';
import createTestStore from '../../../../test/createTestStore';
import { book as archiveBook, page, pageInChapter, pageInOtherChapter } from '../../../../test/mocks/archiveLoader';
import { mockCmsBook } from '../../../../test/mocks/osWebLoader';
import TestContainer from '../../../../test/TestContainer';
import { Store } from '../../../types';
import { assertWindow } from '../../../utils';
import { receiveBook, receivePage } from '../../actions';
import SectionHighlights from '../../components/SectionHighlights';
import allImagesLoaded from '../../components/utils/allImagesLoaded';
import { SummaryHighlights } from '../../highlights/types';
import { getHighlightLocationFilterForPage } from '../../highlights/utils';
import LoaderWrapper from '../../styles/LoaderWrapper';
import { formatBookData } from '../../utils';
import { receiveStudyGuidesTotalCounts, receiveSummaryStudyGuides, setSummaryFilters } from '../actions';
import { studyGuidesLocationFilters } from '../selectors';
import StudyGuides, { NoStudyGuidesTip } from './StudyGuides';

const hlBlue = {
  annotation: 'hl1', color: HighlightColorEnum.Blue,
  highlightedContent: 'content', id: 'hl1', sourceId: 'testbook1-testpage1-uuid',
};
const hlGreen = {
  annotation: 'hl', color: HighlightColorEnum.Green,
  highlightedContent: 'content', id: 'hl2',  sourceId: 'testbook1-testpage1-uuid',
};
const hlPink = {
  annotation: 'hl', color: HighlightColorEnum.Pink,
  highlightedContent: 'content', id: 'hl3', sourceId: 'testbook1-testpage1-uuid',
};
const hlPurple = {
  annotation: 'hl', color: HighlightColorEnum.Purple,
  highlightedContent: 'content', id: 'hl4', sourceId: 'testbook1-testpage1-uuid',
};
const hlYellow = {
  color: HighlightColorEnum.Yellow,
  highlightedContent: 'content', id: 'hl5', sourceId: 'testbook1-testpage1-uuid',
};

describe('StudyGuides', () => {
  const book = formatBookData(archiveBook, mockCmsBook);
  let store: Store;
  let services: ReturnType<typeof createTestServices>;

  beforeEach(() => {
    store = createTestStore();

    store.dispatch(receiveBook(book));
    store.dispatch(receivePage({...page, references: []}));

    services = createTestServices();
  });

  it('properly display summary highlights', () => {
    const state = store.getState();
    const locationFilters = studyGuidesLocationFilters(state);
    const firstLocation = getHighlightLocationFilterForPage(locationFilters, pageInChapter);
    const secondLocation = getHighlightLocationFilterForPage(locationFilters, pageInOtherChapter);

    if (!firstLocation || !secondLocation) {
      return expect([firstLocation, secondLocation]).not.toContain(undefined);
    }

    const summaryHighlights = {
      [firstLocation.id]: {
        [pageInChapter.id]: [
          {...hlBlue, highlightedContent: '<a href="link">Link</a>'},
          hlGreen,
          hlPink,
          hlPurple,
          {...hlYellow, color: null} /* for coverage for css when color was not found */,
        ],
      },
      [secondLocation.id]: {
        [pageInOtherChapter.id]: [hlBlue, hlGreen],
      },
    } as SummaryHighlights;

    store.dispatch(receiveSummaryStudyGuides(summaryHighlights, {pagination: null}));

    const component = renderer.create(<TestContainer services={services} store={store}>
      <StudyGuides />
    </TestContainer>);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('add promises to promise collector at render', () => {
    const container = assertWindow().document.createElement('div');
    const spyPromiseCollectorAdd = jest.spyOn(services.promiseCollector, 'add');

    const locationFilters = studyGuidesLocationFilters(store.getState());
    const firstLocation = getHighlightLocationFilterForPage(locationFilters, pageInChapter);
    const secondLocation = getHighlightLocationFilterForPage(locationFilters, pageInOtherChapter);

    if (!firstLocation || !secondLocation) {
      return expect([firstLocation, secondLocation]).not.toContain(undefined);
    }

    const summaryHighlights = {
      [firstLocation.id]: {
        [pageInChapter.id]: [
          hlBlue,
          hlGreen,
          hlPink,
          hlPurple,
          {...hlYellow, color: null} /* for coverage for css when color was not found */,
        ],
      },
      [secondLocation.id]: {
        [pageInOtherChapter.id]: [hlBlue, hlGreen],
      },
    } as SummaryHighlights;

    store.dispatch(receiveSummaryStudyGuides(summaryHighlights, {pagination: null}));

    renderer.create(<TestContainer services={services} store={store}>
      <StudyGuides />
    </TestContainer>, { createNodeMock: () => container });

    expect(spyPromiseCollectorAdd).toHaveBeenCalledWith(allImagesLoaded(container));
    expect(spyPromiseCollectorAdd).toHaveBeenCalledWith(typesetMath(container, assertWindow()));
  });

  it('show loading state on filters change', () => {
    const state = store.getState();
    const locationFilters = studyGuidesLocationFilters(state);
    const firstLocation = getHighlightLocationFilterForPage(locationFilters, pageInChapter);
    const secondLocation = getHighlightLocationFilterForPage(locationFilters, pageInOtherChapter);

    if (!firstLocation || !secondLocation) {
      return expect([firstLocation, secondLocation]).not.toContain(undefined);
    }

    store.dispatch(receiveStudyGuidesTotalCounts({
      [firstLocation.id]: {[HighlightColorEnum.Green]: 5},
      [secondLocation.id]: {[HighlightColorEnum.Green]: 2},
    }));

    const summaryHighlights = {
      [firstLocation.id]: {
        [pageInOtherChapter.id]: [hlBlue, hlGreen, hlPink, hlPurple, hlYellow],
      },
      [secondLocation.id]: {
        [pageInChapter.id]: [hlBlue, hlGreen],
      },
    } as SummaryHighlights;

    renderer.act(() => {
      store.dispatch(setSummaryFilters({locationIds: [firstLocation.id, secondLocation.id]}));
      store.dispatch(receiveSummaryStudyGuides(summaryHighlights, {pagination: null}));
    });

    const component = renderer.create(<TestContainer services={services} store={store}>
      <StudyGuides/>
    </TestContainer>);

    const sections = component.root.findAllByType(SectionHighlights);
    expect(sections.length).toEqual(2);

    expect(component.root.findAllByType(LoaderWrapper).length).toEqual(0);

    renderer.act(() => {
      store.dispatch(setSummaryFilters({locationIds: [firstLocation.id, secondLocation.id]}));
    });

    const isLoading = component.root.findByType(LoaderWrapper);
    expect(isLoading).toBeDefined();
  });

  it('shows "no results" state', () => {
    const summaryStudyGuides = {} as SummaryHighlights;
    store.dispatch(receiveSummaryStudyGuides(summaryStudyGuides, {pagination: null}));

    const component = renderer.create(<TestContainer services={services} store={store}>
      <StudyGuides />
    </TestContainer>);

    expect(component.root.findByProps(NoStudyGuidesTip)).toBeDefined();
  });
});
