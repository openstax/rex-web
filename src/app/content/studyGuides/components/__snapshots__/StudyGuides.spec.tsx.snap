// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudyGuides properly display summary highlights 1`] = `
<div
  className="StudyGuides-sc-1d60zwp-0 cAwzzu"
>
  <div
    className="HighlightsWrapper-vefi0f-0 kuspas"
  >
    <div
      className="SectionHighlights__HighlightsChapterWrapper-c51y19-0 kARwht"
    >
      <div
        className="SectionHighlights__HighlightsChapter-c51y19-1 HowFS"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-number\\">10</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 5</span>",
          }
        }
        data-testid="chapter-title"
      />
    </div>
    <div
      className="SectionHighlights__HighlightWrapper-c51y19-2 gFcHNj"
    >
      <div
        className="SectionHighlights__HighlightSection-c51y19-3 cYxWVd"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-text\\">Test Page 6</span>",
          }
        }
        data-testid="section-title"
      />
      <div
        className="StudyGuidesListElement__HighlightOuterWrapper-sc-5qgyda-0 cEYBxy"
      >
        <div
          className="StudyGuidesListElement__HighlightAnnotation-sc-5qgyda-1 cZidrZ"
        >
          hl1
        </div>
        <div
          className="StudyGuidesListElement__HighlightContentWrapper-sc-5qgyda-2 lbsMtc"
          color="blue"
        >
          <div
            className="content-excerpt ContentExcerpt-sc-16hg3lt-0 iKlVB sc-bdVaJa ibTDPh"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<a href=\\"/books/book-slug-1/pages/link\\" target=\\"_blank\\">Link</a>",
              }
            }
            data-dynamic-style={false}
            data-highlight-id="hl1"
          />
        </div>
      </div>
      <div
        className="StudyGuidesListElement__HighlightOuterWrapper-sc-5qgyda-0 cEYBxy"
      >
        <div
          className="StudyGuidesListElement__HighlightAnnotation-sc-5qgyda-1 cZidrZ"
        >
          hl
        </div>
        <div
          className="StudyGuidesListElement__HighlightContentWrapper-sc-5qgyda-2 jQASvQ"
          color="green"
        >
          <div
            className="content-excerpt ContentExcerpt-sc-16hg3lt-0 iKlVB sc-bdVaJa ibTDPh"
            dangerouslySetInnerHTML={
              Object {
                "__html": "content",
              }
            }
            data-dynamic-style={false}
            data-highlight-id="hl2"
          />
        </div>
      </div>
      <div
        className="StudyGuidesListElement__HighlightOuterWrapper-sc-5qgyda-0 cEYBxy"
      >
        <div
          className="StudyGuidesListElement__HighlightAnnotation-sc-5qgyda-1 cZidrZ"
        >
          hl
        </div>
        <div
          className="StudyGuidesListElement__HighlightContentWrapper-sc-5qgyda-2 kuhHtd"
          color="pink"
        >
          <div
            className="content-excerpt ContentExcerpt-sc-16hg3lt-0 iKlVB sc-bdVaJa ibTDPh"
            dangerouslySetInnerHTML={
              Object {
                "__html": "content",
              }
            }
            data-dynamic-style={false}
            data-highlight-id="hl3"
          />
        </div>
      </div>
      <div
        className="StudyGuidesListElement__HighlightOuterWrapper-sc-5qgyda-0 cEYBxy"
      >
        <div
          className="StudyGuidesListElement__HighlightAnnotation-sc-5qgyda-1 cZidrZ"
        >
          hl
        </div>
        <div
          className="StudyGuidesListElement__HighlightContentWrapper-sc-5qgyda-2 fMEUWV"
          color="purple"
        >
          <div
            className="content-excerpt ContentExcerpt-sc-16hg3lt-0 iKlVB sc-bdVaJa ibTDPh"
            dangerouslySetInnerHTML={
              Object {
                "__html": "content",
              }
            }
            data-dynamic-style={false}
            data-highlight-id="hl4"
          />
        </div>
      </div>
      <div
        className="StudyGuidesListElement__HighlightOuterWrapper-sc-5qgyda-0 cEYBxy"
      >
        <div
          className="StudyGuidesListElement__HighlightAnnotation-sc-5qgyda-1 cZidrZ"
        />
        <div
          className="StudyGuidesListElement__HighlightContentWrapper-sc-5qgyda-2 cjhgbG"
          color={null}
        >
          <div
            className="content-excerpt ContentExcerpt-sc-16hg3lt-0 iKlVB sc-bdVaJa ibTDPh"
            dangerouslySetInnerHTML={
              Object {
                "__html": "content",
              }
            }
            data-dynamic-style={false}
            data-highlight-id="hl5"
          />
        </div>
      </div>
    </div>
    <div
      className="SectionHighlights__HighlightsChapterWrapper-c51y19-0 kARwht"
    >
      <div
        className="SectionHighlights__HighlightsChapter-c51y19-1 HowFS"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-number\\">12</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 6</span>",
          }
        }
        data-testid="chapter-title"
      />
    </div>
    <div
      className="SectionHighlights__HighlightWrapper-c51y19-2 gFcHNj"
    >
      <div
        className="SectionHighlights__HighlightSection-c51y19-3 cYxWVd"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-text\\">Test Page 7</span>",
          }
        }
        data-testid="section-title"
      />
      <div
        className="StudyGuidesListElement__HighlightOuterWrapper-sc-5qgyda-0 cEYBxy"
      >
        <div
          className="StudyGuidesListElement__HighlightAnnotation-sc-5qgyda-1 cZidrZ"
        >
          hl1
        </div>
        <div
          className="StudyGuidesListElement__HighlightContentWrapper-sc-5qgyda-2 lbsMtc"
          color="blue"
        >
          <div
            className="content-excerpt ContentExcerpt-sc-16hg3lt-0 iKlVB sc-bdVaJa ibTDPh"
            dangerouslySetInnerHTML={
              Object {
                "__html": "content",
              }
            }
            data-dynamic-style={false}
            data-highlight-id="hl1"
          />
        </div>
      </div>
      <div
        className="StudyGuidesListElement__HighlightOuterWrapper-sc-5qgyda-0 cEYBxy"
      >
        <div
          className="StudyGuidesListElement__HighlightAnnotation-sc-5qgyda-1 cZidrZ"
        >
          hl
        </div>
        <div
          className="StudyGuidesListElement__HighlightContentWrapper-sc-5qgyda-2 jQASvQ"
          color="green"
        >
          <div
            className="content-excerpt ContentExcerpt-sc-16hg3lt-0 iKlVB sc-bdVaJa ibTDPh"
            dangerouslySetInnerHTML={
              Object {
                "__html": "content",
              }
            }
            data-dynamic-style={false}
            data-highlight-id="hl2"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
